runtime/onert/backend/train/ops/FullyConnectedLayer.cc

```diff
@@ -0,0 +1,179 @@
+  }
+}
+
+void FullyConnectedLayer::forward(bool) { cpu::ops::FullyConnectedLayer::run(); }
+
+void FullyConnectedLayer::backward(uint32_t)
+{
+  const auto data_type = _deriv_output->data_type();
+  assert(data_type == _input->data_type());
+  switch (data_type)
+  {
+    case OperandType::FLOAT32:
+    {
+      assert(data_type == _grad_weights->data_type());
+      assert(data_type == _grad_bias->data_type());
+      backwardFloat32();
+      break;
+    }
+    default:
+      throw std::runtime_error{"train ConvolutionLayer: unsupported data type"};
+  }
+}
+
+void FullyConnectedLayer::backwardFloat32()
+{
+  // Calculate gradient for activation
+  const IPortableTensor *act_grad;
+  switch (_activation)
+  {
+    case ir::Activation::NONE:
+      act_grad = _deriv_output;
+      break;
+    case ir::Activation::RELU:
+      ReLUGrad(getShape(_output), getBuffer<float>(_output), getShape(_deriv_output),
+               getBuffer<float>(_deriv_output), getShape(_act_deriv_output.get()),
+               getBuffer<float>(_act_deriv_output.get()));
+      act_grad = _act_deriv_output.get();
+      break;
+    default:
+      throw std::runtime_error("train ConvolutionLayer: Unsupported activation type yet");
+      break;
+  }
+
+  // Initialize TransposeParams
+  nnfw::cker::TransposeParams transpose_param;
+  transpose_param.perm_count = 2;
+  transpose_param.perm[0] = 1;
+  transpose_param.perm[1] = 0;
+
+  // Initialize FullyConnectedParams
+  nnfw::cker::FullyConnectedParams op_params;
+  op_params.activation = nnfw::cker::FusedActivationFunctionType::kNone;
+
+  // Transpose and compute gradient for input
+  auto transposed_weights = _transposed_weights.get();
+  assert(transposed_weights->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(_weights), getBuffer<float>(_weights),
+                        getShape(transposed_weights), getBuffer<float>(transposed_weights));
+
+  nnfw::cker::FullyConnected(op_params, getShape(act_grad), getBuffer<float>(act_grad),
+                             getShape(transposed_weights), getBuffer<float>(transposed_weights),
+                             getShape(nullptr), nullptr, getShape(_deriv_input),
+                             getBuffer<float>(_deriv_input));
+
+  // Transpose and compute gradient for weights
+  auto transposed_input = _transposed_input.get();
+  assert(transposed_input->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(_input), getBuffer<float>(_input),
+                        getShape(transposed_input), getBuffer<float>(transposed_input));
+
+  auto transposed_deriv_output = _transposed_deriv_output.get();
+  assert(transposed_deriv_output->getShape().rank() == 2);
+  nnfw::cker::Transpose(transpose_param, getShape(act_grad), getBuffer<float>(act_grad),
+                        getShape(transposed_deriv_output),
+                        getBuffer<float>(transposed_deriv_output));
```

Comment:
(Q) Why do you transpose `deriv_output`?

refer: https://www.adityaagrawal.net/blog/deep_learning/bprop_fc

Reply:
Considering the shapes. transpose is required to use with fc kernel.

output {2, 3} = fc( input {2, 5}, weights {3, 5} )
∂L/∂X {2,5} = fc( incoming gradient {2, 3}, transposed weights {5, 3} )
∂L/∂W {3, 5} = transpose ( fc( transposed input {5, 2}, transposed incoming gradient {3, 2} ) )

I haven't checked the formula in detail. So I will check this implementation with the formula.

Reply:
I confirmed the formula in detail. 

$fc (X, W^T) = matmul(X, W)$


output {2, 3} = fc( input {2, 5}, weights {3, 5} )$
∂L/∂X {2, 5} = fc( incoming gradient {2, 3}, transposed weights {5, 3} )
∂L/∂W {3, 5} = fc( transposed incoming gradient {3, 2},  transposed input {5, 2} )

output {2, 3} = matmul( input {2, 5}, transposed weights {5, 3} )
∂L/∂X {2, 5} = matmul( incoming gradient {2, 3}, weights {3, 5} )
∂L/∂W {3, 5} = matmul( transposed incoming gradient {3, 2},  input {2, 5} )

If $W$ is transposed weight, input is $X$, and output is $Z$,
$Z = matmul( X, W )$
$∂L/∂X = matmul( ∂L/∂Z, W^T )$
$∂L/∂W = matmul( (∂L/∂Z)^T , I )$

Reply:
Could you let me know a link to the weight derivative information?

I found the following formula in your repo.

> #### Weights
> $\frac{\partial L}{\partial W} = (\frac{\partial L}{\partial Z})^T X^T$

But it's different from my formula. So I want to know the details.

> ### $\cfrac{\partial L}{\partial W} = \cfrac{\partial L}{\partial Z} X^{T}$

Reply:
The formulas in the link are based on the forward $Z = WX + B$
My example formulas are based on the forward $Z = XW + B$ 

Matrix multiplication is not commutative, so we cannot use that formula in the link as it is.
For example, since $(AB)^T = B^TA^T$ and $AB = ((AB)^T)^T$, the above formula ($Z = WX + B$) is changed to $Z = (X^T W^T)^T + B$ so that the results of Z are the same.

If you think of $X$ as $W$ and $X$ as $W$, the formula I wrote and the formula in the link are the same.
$Z = matmul( W,  X)$
$∂L/∂X = matmul( (∂L/∂Z)^T , W )$
$∂L/∂W = matmul( ∂L/∂Z, X^T )$